client/allocrunner/alloc_runner_linux.go

```diff
@@ -0,0 +1,108 @@
+	ns := &allocNetworkIsolationSetter{ar: ar}
+	nm, err := ar.initNetworkManager()
+	if err != nil {
+		return err
+	}
+
+	ar.runnerHooks = append(ar.runnerHooks, newNetworkHook(ns, hookLogger, ar.Alloc(), nm))
+	return nil
+}
+
+func (ar *allocRunner) initNetworkManager() (nm drivers.DriverNetworkManager, err error) {
+	// The defaultNetworkManager is used if a driver doesn't need to create the network
+	nm = &defaultNetworkManager{}
+	tg := ar.Alloc().Job.LookupTaskGroup(ar.Alloc().TaskGroup)
+
+	// default netmode to host, this can be overridden by the task or task group
+	tgNetMode := "host"
+	if len(tg.Networks) > 0 && tg.Networks[0].Mode != "" {
+		tgNetMode = tg.Networks[0].Mode
+	}
+
+	// networkInitiator tracks the task driver which needs to create the network
+	// to check for multiple drivers needing the create the network
+	var networkInitiator string
+
+	// driverCaps tracks which drivers we've checked capabilities for so as not
+	// to do extra work
+	driverCaps := make(map[string]struct{})
+	for _, task := range tg.Tasks {
+		// the task's netmode defaults to the the task group but can be overridden
+		taskNetMode := tgNetMode
+		if len(task.Resources.Networks) > 0 && task.Resources.Networks[0].Mode != "" {
+			taskNetMode = task.Resources.Networks[0].Mode
+		}
+
+		// netmode host should always work to support backwards compat
+		if taskNetMode == "host" {
+			continue
+		}
+
+		// check to see if capabilities of this task's driver have already been checked
+		if _, ok := driverCaps[task.Driver]; ok {
+			continue
+		}
+
+		driver, err := ar.driverManager.Dispense(task.Driver)
+		if err != nil {
+			return nil, fmt.Errorf("failed to dispense driver %s: %v", task.Driver, err)
+		}
+
+		caps, err := driver.Capabilities()
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrive capabilities for driver %s: %v",
+				task.Driver, err)
+		}
+
+		// check that the driver supports the requested network isolation mode
+		netIsolationMode := netModeToIsolationMode(taskNetMode)
+		if !caps.HasNetIsolationMode(netIsolationMode) {
+			return nil, fmt.Errorf("task %s does not support %q networking mode", task.Name, taskNetMode)
+		}
+
+		// check if the driver needs to create the network and if a different
+		// driver has already claimed it needs to initiate the network
+		if caps.MustInitiateNetwork {
+			if networkInitiator != "" {
+				return nil, fmt.Errorf("tasks %s and %s want to initiate networking but only one is able", networkInitiator, task.Name)
+			}
+
+			nm = driver
+			networkInitiator = task.Name
+		}
+
+		// mark this driver's capabilities as checked
+		driverCaps[task.Driver] = struct{}{}
+	}
+
+	return nm, nil
+}
+
+func netModeToIsolationMode(netMode string) drivers.NetIsolationMode {
+	switch strings.ToLower(netMode) {
+	case "host":
+		return drivers.NetIsolationModeHost
+	case "bridge", "none":
+		return drivers.NetIsolationModeGroup
+	case "driver":
+		return drivers.NetIsolationModeTask
```

Comment:
I guess these are defined in another PR - but is this a hard-coded list and should we make job resources network mode use task/group/host strings instead?

The reason I'm asking is how this is related to potentially other networking modes, e.g. `macvlan`, `weaves`, etc and how we envision extensibility of network types.

Reply:
Other networking modes will be supported through CNI. Likely the network name for these will be `cni/<network-name>` and this logic will be updated then.

I'd argue against mixing the task/group/host (network isolation mode) naming as they have a slightly different purpose. The net isolation mode is for the driver's perspective and doesn't carry any meaning about the network. Where is the network mode would have deeper meaning, especially once CNI is supported.

Reply:
Ah - makes sense now.  all good then.